### PR TITLE
CMake tutorial

### PR DESCRIPTION
**Pull request by TylerMSFT**

### CODE REVIEW

**Comment by TylerMSFT** [comment]
#hold-off

**Comment by jureid** [comment]
Wow, thanks Tyler! This looks great! A few suggestions:

1.  In the "Create a CMake quickstart" section, can we include a sentence that says something like: _If you have an existing CMake project (with CMakeLists.txt in the project root directory), you can skip to the next section to learn how to configure your project._ We think that most customers will use CMake Tools to configure an existing CMake project rather than use the Quickstart functionality.

2. Should "intellisense" be capitalized as "IntelliSense"? Not sure what the rule of thumb is here

3. Is the first "Select a kit" section supposed to be a subheading of "Create a CMake quickstart"? It makes sense to me for it to be a subheading, since they have to select a kit as part of the quickstart process.

4. In the very latest release of CMake Tools, the extension will automatically scan for kits the first time the customer runs the "Select a kit" command. I suggest rewording that part to: _1. Open the Command Palette (kb(workbench.action.showCommands)) and run CMake: Select a Kit. The CMake Tools extension will automatically scan for kits on your computer, creating a list of compilers found on your system. 2. Select the compiler you want to use. For example, depending on the compilers you have installed, you might see something like:_

5. CMakeKits.json is actually called cmake-tools-kits.json (you can search for it by doing ctrl, P)

File: docs/cpp/cmake-linux.md [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+1. [Visual Studio Code](/download).
+1. [C++ extension for VS Code](https://marketplace.visualstudio.com/items?itemName=ms-vscode.cpptools). You can install the C/C++ extension by searching for 'c++' in the Extensions view (`kb(workbench.view.extensions)`).
+
+    ![C/C++ extension](images/cpp/cpp-extension.png)
+
+1. [CMake Tools extension for VS Code](https://marketplace.visualstudio.com/items?itemName=ms-vscode.cmake-tools). You can install the C/C++ extension by searching for 'cmake tools' in the Extensions view (`kb(workbench.view.extensions)`).
+
+    ![CMake tools extension](images/cpp/cmake-extension.png)
+
+1. You'll also need to install CMake, a compiler, and a debugger.
```
Comment (esweet431) [inline docs/cpp/cmake-linux.md:33]: When you install build-essential you are also installing make, which is the generator that CMake will use. You don't need to add any additional steps, but maybe explicitly call out that you need a generator, and highlight that build-essential installs both GNU compiler tools and make?

File: docs/cpp/cmake-linux.md [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+Area: cpp
+TOCTitle: Get started with CMake Tools on Linux
+ContentId: 86543311-5452-4b1f-a44c-03cc3df04c3f
+PageTitle: Get started with CMake Tools on Linux 
+DateApproved: 06/11/2020
+MetaDescription: Get started with the CMake Tools Visual Studio Code extension on Linux 
+---
+# Get started with CMake Tools on Linux
+
+CMake is an open-source, cross-platform tool for building, testing, and packaging software. It uses platform and compiler independent configuration files to generate native makefiles that can be used with your compiler, on your platform.
```
Comment (esweet431) [inline docs/cpp/cmake-linux.md:12]: Nit: Makefiles is just one build system that CMake supports. Maybe 'generate native build tool files'?

Comment (TylerMSFT) [reply docs/cpp/cmake-linux.md:12]: fixed in next commit.

File: docs/cpp/cmake-linux.md (hunk 1 above)
Comment (TylerMSFT) [reply docs/cpp/cmake-linux.md:33]: Done.

File: docs/cpp/cmake-linux.md [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+1. [Visual Studio Code](/download).
+1. [C++ extension for VS Code](https://marketplace.visualstudio.com/items?itemName=ms-vscode.cpptools). You can install the C/C++ extension by searching for 'c++' in the Extensions view (`kb(workbench.view.extensions)`).
+
+    ![C/C++ extension](images/cpp/cpp-extension.png)
+
+1. [CMake Tools extension for VS Code](https://marketplace.visualstudio.com/items?itemName=ms-vscode.cmake-tools). You can install the C/C++ extension by searching for 'cmake tools' in the Extensions view (`kb(workbench.view.extensions)`).
+
+    ![CMake tools extension](images/cpp/cmake-extension.png)
+
+1. You'll also need to install CMake, a compiler, a debugger, and .
```
Comment (esweet431) [inline docs/cpp/cmake-linux.md:33]: 'and .'
Maybe this was supposed to be 'and a generator'?

File: docs/cpp/cmake-linux.md [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+cmake --version
+```
+
+To install, or get the latest version of CMake on your system, see the instructions for your platform on the [Kitware APT Repository](https://apt.kitware.com/). You want to install a version greater than 3.17.
+
+### Ensure that development tools are installed
+
+Although you'll use VS Code to edit your source code, you'll compile and debug the source code using the compiler, debugger, and build tools (such as make) installed on your system.
+
+For this tutorial on Ubuntu, we'll use the GCC compiler, GDB to debug, and related development tools. These tools are not installed by default on Ubuntu, so you need to install them. Fortunately, that's easy.
```
Comment (esweet431) [inline docs/cpp/cmake-linux.md:51]: Suggestion: 'GDB to debug, and make to generate native build files.'

File: docs/cpp/cmake-linux.md [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+gcc -v
+```
+
+If GCC isn't installed, run the following command from the terminal window to update the Ubuntu package lists. An out-of-date Linux distribution can interfere with getting the latest packages.
+
+```bash
+sudo apt-get update
+```
+
+Next install the GNU compiler, development tools, and GDB debugger with this command:
```
Comment (esweet431) [inline docs/cpp/cmake-linux.md:67]: Suggestion: '...the GNU compiler, make, and the GDB debugger...'

File: docs/cpp/cmake-linux.md (hunk 3 above)
Comment (TylerMSFT) [reply docs/cpp/cmake-linux.md:33]: fixed :-)

File: docs/cpp/cmake-linux.md (hunk 4 above)
Comment (TylerMSFT) [reply docs/cpp/cmake-linux.md:51]: fixed

File: docs/cpp/cmake-linux.md (hunk 5 above)
Comment (TylerMSFT) [reply docs/cpp/cmake-linux.md:67]: fixed

File: docs/cpp/cmake-linux.md [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+ContentId: 86543311-5452-4b1f-a44c-03cc3df04c3f
+PageTitle: Get started with CMake Tools on Linux 
+DateApproved: 06/11/2020
+MetaDescription: Get started with the CMake Tools Visual Studio Code extension on Linux 
+---
+# Get started with CMake Tools on Linux
+
+CMake is an open-source, cross-platform tool for building, testing, and packaging software. It uses platform and compiler independent configuration files to generate native build tool files specific to your compiler and platform.
+
+The CMake Tools extension integrates Visual Studio Code and CMake to make it easy to configure, build, and debug your C++ project using CMake.
```
Comment (bobbrow) [inline docs/cpp/cmake-linux.md:14]: The "using CMake" at the end of sentence reads a bit redundant to me.

File: docs/cpp/cmake-linux.md [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+
+1. [CMake Tools extension for VS Code](https://marketplace.visualstudio.com/items?itemName=ms-vscode.cmake-tools). You can install the C/C++ extension by searching for 'cmake tools' in the Extensions view (`kb(workbench.view.extensions)`).
+
+    ![CMake tools extension](images/cpp/cmake-extension.png)
+
+1. You'll also need to install CMake, a compiler, a debugger, and a generator.
+
+### Ensure that CMake is installed
+
+The VS Code CMake Tools extension does its work using CMake installed on your system. For best results, use CMake version 3.17 or greater.
```
Comment (bobbrow) [inline docs/cpp/cmake-linux.md:37]: I was thinking 3.15 was the min version we recommended. 3.17 is currently the latest version, but it will be outdated in the coming months and years, so it might be fine to leave it.

File: docs/cpp/cmake-linux.md [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+
+See if CMake is already installed on your system. Open a Terminal window and enter the following command:
+
+```bash
+cmake --version
+```
+
+To install, or get the latest version of CMake on your system, see the instructions for your platform on the [Kitware APT Repository](https://apt.kitware.com/). You want to install a version greater than 3.17.
```
Comment (bobbrow) [inline docs/cpp/cmake-linux.md:45]: I don't think the second sentence is necessary.  If they install from the APT repo, they will get the latest cmake.

File: docs/cpp/cmake-linux.md [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+sudo apt-get update
+```
+
+Next, install the GNU compiler, make, and the GDB debugger with this command:
+
+```bash
+sudo apt-get install build-essential gdb
+```
+
+## Create a CMake quickstart
```
Comment (bobbrow) [inline docs/cpp/cmake-linux.md:73]: This reads awkwardly to me.  I expected it to read like "Quick start: Create a CMake project" or just "Create a CMake project".

File: docs/cpp/cmake-linux.md [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+
+**Note:** If you had wanted to create a basic source and header file, you would have selected **Library** instead. But for this tutorial, **Executable** will do. If you are prompted to configure IntelliSense for the folder, you can either ignore the prompt or select **Allow**.
+
+This creates a hello world CMake project containing `main.cpp`, `CMakeLists.txt` (which tells the CMake tools how to build your project), and a folder named **build** for your build files:
+
+![Project contents](images/cpp/cmake-project-contents.png)
+
+### Select a kit
+
+Before you can use the CMake tools extension to build a project, you need to configure it to know about the compilers on your system. Do that by scanning for what CMake calls 'kits'. A kit represents a toolchain, which is the compilers, linkers, and other tools used to build your project. To scan for kits:
```
Comment (bobbrow) [inline docs/cpp/cmake-linux.md:107]: CMake Tools ("Tools" with a capital "T").
`CMake calls 'kits'` --> `CMake Tools calls 'kits'`  (CMake itself doesn't use this terminology)

File: docs/cpp/cmake-linux.md [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+
+![Create CMake quickstart](images/cpp/cmake-quickstart-command-palette.png)
+
+Enter a project name. This will be written to `CMakeLists.txt` and a few initial source files.
+
+Next, select **Executable** as the project type to create a basic source file (`main.cpp`) that includes a basic `main()` function.
+
+![Choose project type](images/cpp/cmake-choose-type.png)
+
+**Note:** If you had wanted to create a basic source and header file, you would have selected **Library** instead. But for this tutorial, **Executable** will do. If you are prompted to configure IntelliSense for the folder, you can either ignore the prompt or select **Allow**.
```
Comment (bobbrow) [inline docs/cpp/cmake-linux.md:99]: People should not ignore the prompt. They will only get the best IntelliSense experience if they click the "Allow" button.

File: docs/cpp/cmake-linux.md [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+
+## Configure Hello World
+
+There are two things you must do to configure your CMake project: select a kit (which you just did) and select a variant.  
+
+The kit you selected previously is shown in the status bar. For example:
+
+![Selected kit in status bar](images/cpp/cmake-kit-statusbar.png)
+
+To change the kit, you can click on the kit in the status bar, or run the **CMake: Select a kit** again from the Command Palette. If you don't see the compiler you're looking for, you can edit the `cmake-tools-kits.json` file in your project and add it manually.
```
Comment (bobbrow) [inline docs/cpp/cmake-linux.md:123]: Does `cmake-tools-kits.json` get created automatically?  I didn't think so.  People might now know what this file is.  They can run the "Edit user-local CMake Kits" command to add the kit to the global list or create a `cmake-tools-kits.json` in the .vscode folder if they want to create a project-specific kit.  I'm not sure how much detail you want to go into here though.

File: docs/cpp/cmake-linux.md [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+
+The selected variant will appear in the status bar next to the active kit.  
+
+### CMake: Configure
+
+Now that you've selected a kit and a variant, open the Command Palette (`kb(workbench.action.showCommands)`) and run the **CMake: Configure** command to configure your project. This generates build files in the project's build folder using the kit and variant you selected.
+
+## Build hello world
+
+After configuring your project, you're ready to build. Open the Command Palette (`kb(workbench.action.showCommands)`) and run the **CMake: Build**, or simply select the **Build** button from the status bar.
```
Comment (bobbrow) [inline docs/cpp/cmake-linux.md:150]: missing the word "command" after the command.

File: docs/cpp/cmake-linux.md [hunk 14]
```diff
@@ -0,0 +1,176 @@
...
+![Build](images/cpp/cmake-build.png)
+
+You can select which targets you'd like to build by selecting **CMake: Set Build Target** from the Command Palette. 
+
+By default, CMake Tools builds all targets. The selected target will appear in the status bar next to the **Build** button.
+
+## Debug hello world
+
+To run and debug your project, open `main.cpp` and put a breakpoint on the `std::cout` line.
+Then open the Command Palette (`kb(workbench.action.showCommands)`) and run **CMake: Debug**. The debugger  stops on the `std::out` line:
```
Comment (bobbrow) [inline docs/cpp/cmake-linux.md:161]: "c" missing in `std::cout`

File: docs/cpp/cmake-linux.md (hunk 6 above)
Comment (TylerMSFT) [reply docs/cpp/cmake-linux.md:14]: Yeah - fixed.

File: docs/cpp/cmake-linux.md (hunk 7 above)
Comment (TylerMSFT) [reply docs/cpp/cmake-linux.md:37]: Set to 3.15

Comment (TylerMSFT) [reply docs/cpp/cmake-linux.md:37]: Set to 3.15

File: docs/cpp/cmake-linux.md (hunk 8 above)
Comment (TylerMSFT) [reply docs/cpp/cmake-linux.md:45]: This instruction is for people who have already installed it but don't have the latest version.

File: docs/cpp/cmake-linux.md (hunk 9 above)
Comment (TylerMSFT) [reply docs/cpp/cmake-linux.md:73]: Not my wording. Changed to 'Create a CMake project'

File: docs/cpp/cmake-linux.md (hunk 10 above)
Comment (TylerMSFT) [reply docs/cpp/cmake-linux.md:107]: Fixed. I think I have them all.

Comment (TylerMSFT) [reply docs/cpp/cmake-linux.md:107]: Fixed. I think I have them all.

File: docs/cpp/cmake-linux.md (hunk 13 above)
Comment (TylerMSFT) [reply docs/cpp/cmake-linux.md:150]: fixed that one. Plus another that was like it.

File: docs/cpp/cmake-linux.md (hunk 14 above)
Comment (TylerMSFT) [reply docs/cpp/cmake-linux.md:161]: fixed

File: docs/cpp/cmake-linux.md (hunk 10 above)
Comment (TylerMSFT) [reply docs/cpp/cmake-linux.md:107]: removed "cmake tools calls" and just say "scanning for 'kits'", since this is what the extension calls them.

File: docs/cpp/cmake-linux.md (hunk 11 above)
Comment (TylerMSFT) [reply docs/cpp/cmake-linux.md:99]: fixed

File: docs/cpp/cmake-linux.md (hunk 12 above)
Comment (TylerMSFT) [reply docs/cpp/cmake-linux.md:123]: I added the step to run the cmake:edit user-local cmake kits command.

**Comment by TylerMSFT** [comment]
#sign-off

**Comment by TylerMSFT** [comment]
#sign-off

**Comment by gregvanl** [comment]
Published
